### PR TITLE
WIP a prototype crate design for ops

### PR DESCRIPTION
**Pull request by ry**
cc @afinch7 @bartlomieju 

browse files at https://github.com/ry/deno/tree/op_hello/op_hello

super minimal MVP would be pure JS, https://github.com/ry/deno/tree/op_hello/op_hello_js

This isn't working, just trying to imagine what the ideal structure would look like.

### CODE REVIEW

File: op_hello/src/hello_test.ts [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+//
+// Here it is referencing files across crate boundaries, which will break
+// 'cargo package' and means the crate is not useable outside the deno tree.
+// This might be okay for a first pass, but it's not the best solution.
+//
+// Attempt two:
+// we invent a new URL for referencing files in other crates.
+// this is magic and not browser compatible.. Browser compatibility for
+// ops is not so important.
+import { test } from "crate://deno_std@0.19.0/testing/mod.ts";
```
Comment (ry) [inline op_hello/src/hello_test.ts]: @afinch7 is this along the lines of what you were thinking?

File: op_hello/src/hello_test.ts [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+// Attempt two:
+// we invent a new URL for referencing files in other crates.
+// this is magic and not browser compatible.. Browser compatibility for
+// ops is not so important.
+//
+//  import { test } from "crate://deno_std@0.19.0/testing/mod.ts";
+//
+// This is quite nice. But the version of deno_std already specified in
+// Cargo.toml. I think we shouldn't repeat it.
+import { test } from "crate://deno_std/testing/mod.ts";
```
Comment (bartlomieju) [inline op_hello/src/hello_test.ts]: > // we invent a new URL for referencing files in other crates.
// this is magic and not browser compatible.. Browser compatibility for
// ops is not so important.
//
//  import { test } from "crate://deno_std@0.19.0/testing/mod.ts";
//
// This is quite nice. But the version of deno_std already specified in
// Cargo.toml. I think we shouldn't repeat it.

This is least magical that we can hope for and it's an URL after all. I like this idea very much :+1:

File: op_hello/src/hello.ts [hunk 3]
```diff
@@ -0,0 +1,12 @@
+const OP_HELLO: number = Deno.ops.add("hello");
```
Comment (bartlomieju) [inline op_hello/src/hello.ts]: So this is cool and I'd really like to use it that way, however due to quirks related to `shared_queue` I was unable to make this solution work - [see comments](https://github.com/denoland/deno/pull/3002#discussion_r326865393)

Comment (bartlomieju) [reply op_hello/src/hello.ts]: Actually I'm gonna try to implement it once more, I'll get back to you

Comment (ry) [reply op_hello/src/hello.ts]: Ah, ok - well let's not worry about the exact spelling of this - we can always adjust it in the future. How about this API: 
```js
export function hello() {
  Deno.core.send(Deno.ops["hello"]);
}
```
https://github.com/ry/deno/blob/op_hello/op_hello_js/hello.js

Comment (bartlomieju) [reply op_hello/src/hello.ts]: @ry that will work (given that we "synchronize" ops on runtime initialization), however presented solution brings back @afinch7's concerns about performance hit caused by object lookup on each op call. Still we can start with this API and then figure out how to make it fast

Comment (ry) [reply op_hello/src/hello.ts]: I'm quite sure we're not at the level of perf where a hash look up like that would effect our performance. I hope we are some day...

**Comment by bartlomieju** [comment]
@ry could we game out example with "json" dispatch in this PR as well? It's gonna be most often used method to write ops so it'd be nice to see how we can fit it in this model

**Comment by ry** [comment]
@bartlomieju I'd rather land an example crate that uses core only first. I think it keeps us honest about the interfaces. There are bigger problems of how to get ThreadSafeState into the ops - but it can be solved maybe by introducing a special type of init function which receives a deno_cli::Worker instead of deno::Isolate. We want to still enable some typescript hater to build a "deno.js" style project. So let's make this work in the simple case first in core, then later build on that structure to deliver cli op crates.

**Comment by bartlomieju** [comment]
> There are bigger problems of how to get ThreadSafeState into the ops - but it can be solved maybe by introducing a special type of init function which receives a deno_cli::Worker instead of deno::Isolate. 

I figured it out in #3004, please see `cli/worker.rs`. That solution can be generalized to provide any additional arguments to ops as long as trait boundaries are satisfied.

> We want to still enable some typescript hater to build a "deno.js" style project. So let's make this work in the simple case first in core, then later build on that structure to deliver cli op crates.

That's fine for me, I can tinker on my own

File: op_hello/src/hello.ts (hunk 3 above)
Comment (bartlomieju) [reply op_hello/src/hello.ts]: Should I update #3002 to use this API?

Comment (ry) [reply op_hello/src/hello.ts]: If you're in agreement, yea

Comment (bartlomieju) [reply op_hello/src/hello.ts]: @ry PR updated

File: op_hello_js/lib.rs [hunk 4]
```diff
@@ -0,0 +1,27 @@
+extern crate deno;
+use deno::*;
+
+pub fn init(&mut isolate: Isolate) -> Result<(), ErrBox> {
+  isolate.register_op("hello", op_hello); // register_op defined by #3002
+  isolate.execute("hello.js")?;
```
Comment (bartlomieju) [inline op_hello_js/lib.rs]: I think it should be:
```rust
isolate.execute("hello.js", include_str!("hello.js"));`
```

File: op_hello/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+  //
+  //   import { test } from "crate://deno_std/testing/mod.ts";
+  //
+  // In the future it might make sense to automate this function away, but I think
+  // it would be prudent to make the crate URL resolution as obvious as
+  // possible.
+  isolate.register_crate_url("deno_std", deno_std::get_file);
+
+  // TODO The ability to run typescript doesn't exist in deno core.
+  isolate.run("src/hello.ts")
```
Comment (bartlomieju) [inline op_hello/src/lib.rs]: What about ES modules? AFAIK `Isolate` has only low level API for them and all the complexity is hidden in `modules.rs`.

File: op_hello_js/lib.rs (hunk 4 above)
Comment (ry) [reply op_hello_js/lib.rs]: Yeah, probably. I'm a bit concerned about double counting the source code if we do snapshots...

File: op_hello/src/lib.rs (hunk 5 above)
Comment (ry) [reply op_hello/src/lib.rs]: Yes we should support that too.

File: op_hello/src/hello_test.ts (hunk 1 above)
Comment (afinch7) [reply op_hello/src/hello_test.ts]: This looks good. It's very nice that all the relevant info is encoded into standard elements of the url structure. No need to design fancy parsing.

File: op_hello/src/hello.ts (hunk 3 above)
Comment (afinch7) [reply op_hello/src/hello.ts]: In hindsight I think that hash lookup is fine, and I think my testing was flawed when I originally tested performance with a similar api.

**Comment by ry** [comment]
Closing because this isn't intended to be landed. Just an example branch.

**Pull request closed without merging**